### DISCUSSION TITLE
Need tutorial for Nginx/PHP-FPM users

### DISCUSSION THREAD

**Discussion by manhthang2504** [php/frankenphp#417] [Q&A]
I'm interested in Franken PHP. Read the news, docs... but still confused about how to config it to migrate from Nginx/PHP-FPM.
Apart from benefit, most of the examples show it run one single application, just like `php artisan serve`.

In my case, I have multiple websites on a VPS, NOT using Docker (I'm not sure to call my setup as "bare metal", because from what I understand it may refer to a physical server, but in my case it's VPS). With Nginx, I forward each domain to a PHP-FPM socket. Each PHP-FPM socket run it own linux user (web1_user, web2_user...) (pooling) and my app is chown to that user, to prevent local exploit if one is compromised.

I read article about Roadrunner, Swoole and Franken PHP, seem like there's a new term (to me) called "application server". To my understanding as of now, it mean that one server serve one application, separately. It make me wonder whether it will make conflict in the PORT using. I'm not sure if using Docker can solve it, but I'm not into using Docker in my case. So every app must config a separate port? How to config domain to point to each app? Does FrankenPHP run on shared linux user, or can I config each app running on separate linux user like PHP-FPM pooling above?

Each web app run its own FrankenPHP instance, does it mean that I have several Caddy webserver instances running? Does it create more overhead in compare with one single Nginx service?

Another point is WAF. With Nginx, I have Mod_Security. I know Caddy have Coraza, but there're only tutorial for config Caddy with Coraza, not the Franken PHP itself. From what I read, Franken PHP has Caddy under its hood, but how to config Franken PHP's Caddy? The configuration is same with plain/stand alone Caddy or not? 

If the Franken PHP documentation has the Production Deployment with these explaination for Nginx/PHP-FPM user like me, it will smooth the transition alot.

**Comment by nickchomey** [comment]
I would also suggest some documentation about moving from Apache/Litespeed To Caddy/Frankenphp.

**Comment by nikspyratos** [comment]
I think most of these questions can be answered by playing around with Caddy on its own first to figure out how it usually works - to me at least it reads more like being new to Caddy more than FrankenPHP itself.

>Each web app run its own FrankenPHP instance, does it mean that I have several Caddy webserver instances running?

This one I'm not sure on. The answer to this would likely also relate to the port conflict question - if multiple Caddy instances are trying to use the same port then yeah you'll probably have issues. But if it's a single Caddy instance then it's the same as nginx serving multiple site configs.

**Comment by bsienn-khan** [comment]
Welp all good questions, but no clear answer yet. I guess only way to find out is experiment a lot. Which is really bad to do `Experiements` in production-grade app deployment.

**Comment by mvuidev** [comment]
I'm also interested.

It looks perfect to use in containers, but for a "normal" setup (caddy installed with apt + php-fpm) how to replace caddy with Frankenphp ?

Is it only possible ?

**Comment by mvuidev** [comment]
Ok got it. Just need to use Frankenphp binary globally instead of Caddy, add the "frankenphp"  and "php_server" directives in the Caddy config files, and start Frankenphp like Caddy with --config <path>.

**Reply by nickchomey** [reply]
Would you mind sharing a sample caddyfile?

**Reply by mvuidev** [reply]
Sure :

```
# In my global caddyfile
{
    frankenphp
    order php_server before file_server
}

# In my project caddyfile
domain.com {
    root * /path/to/project/dir
    encode zstd br gzip
    php_server

    # Needed for Wordpress
    try_files {path} {path}/ /index.php?{query}
}
```

Does that help ?

**Reply by nickchomey** [reply]
I think so. I'll have to try it out later. Thanks!

**Comment by gudorian** [comment]
@manhthang2504 
Having had similar concerns as you describe, this is what I ended up doing, maybe it might be of help for your situation as well.

## Replacing PHP-FPM with FrankenPHP in Nginx

If you already know and understand Nginx, you can still use it with FrankenPHP just replacing PHP-FPM very similar to how you likely already do by configuring it as a reverse proxy in Nginx. 
```nginx
map $http_upgrade $connection_upgrade {
    default upgrade;
    ''      close;
}
 
server {
    #...
 
    location /index.php {
        try_files /not_exists @octane;
    }
 
    location / {
        try_files $uri $uri/ @octane;
    }
 
    #...
 
    location @octane {
        set $suffix "";
 
        if ($uri = /index.php) {
            set $suffix ?$query_string;
        }
 
        proxy_http_version 1.1;
        proxy_set_header Host $http_host;
        proxy_set_header Scheme $scheme;
        proxy_set_header SERVER_PORT $server_port;
        proxy_set_header REMOTE_ADDR $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade;
 
        proxy_pass http://127.0.0.1:8000$suffix;
    }
}
```
In the [Laravel Octane Documentation](https://laravel.com/docs/11.x/octane#serving-your-application-via-nginx) you can find the full example on how to do this.
_(Do note though that depending on what is in your apps `public` dir and how nginx is configured that you make sure / verify that ie. `public/some_php_script.php` can't be downloaded via `https://you-app.internal/some_php_script.php`, if you can configure nginx to deny such requests)_

## Serve static files from Nginx
Also, if you still want to serve static assets via Nginx instead of Caddy you can just turn off the file_server directive in your `Caddyfile` (https://frankenphp.dev/docs/performance/#file_server) and assets will be served by Nginx.
```caddy
php_server {
    file_server off
}
```

## Configure PHP
For your PHP (`php.ini`) configuration you can see https://frankenphp.dev/docs/config/ . Not sure if it's the same depending on how you run FrankenPHP, but using the static-build you can just put the php.ini file in the root of your app source dir (same path where you usually would have your `composer.json` file.

## Performance
I've only compared local performance using Caddy directly vs with Nginx proxying Caddy and performance has been comparable (with HTTP/2), though both is a huge improvement from Nginx+PHP-FPM.

And just for full disclosure, I'm running FrankenPHP with HTTPS, but it could be  [quite annoying to configure](https://frankenphp.dev/docs/known-issues/#using-https127001-with-docker), but that enables HTTP/2 and HTTP/3 in Caddy which might have had if different result in performance is observed.
